src_c/transform.c

```diff
@@ -570,6 +570,96 @@ scale_to(pgSurfaceObject *srcobj, pgSurfaceObject *dstobj, int width,
     return retsurf;
 }
 
+SDL_Surface *
+grayscale(pgSurfaceObject *srcobj, pgSurfaceObject *dstobj)
+{
+    SDL_Surface *src = pgSurface_AsSurface(srcobj);
+    SDL_Surface *newsurf;
+
+    if (!dstobj) {
+        newsurf = newsurf_fromsurf(src, srcobj->surf->w, srcobj->surf->h);
+        if (!newsurf)
+            return NULL;
+    }
+    else {
+        newsurf = pgSurface_AsSurface(dstobj);
+    }
+
+    if (newsurf->w != src->w || newsurf->h != src->h) {
+        return (SDL_Surface *)(RAISE(
+            PyExc_ValueError,
+            "Destination surface must be the same size as source surface."));
+    }
+
+    if (src->format->BytesPerPixel != newsurf->format->BytesPerPixel) {
+        return (SDL_Surface *)(RAISE(
+            PyExc_ValueError,
+            "Source and destination surfaces need the same format."));
+    }
+
+    Uint32 *pixels = (Uint32 *)src->pixels;
+
+    SDL_LockSurface(newsurf);
+
+    int i;
+    for (i = 0; i < src->w * src->h; i++) {
```

Comment:
This assumes every pixel is laid out one after the other directly in memory. A->B->C->D

Some surfaces are not this way, and each row of pixels is aligned on some boundary (like a multiple of 64, for example).
So all surfaces have a "stride/pitch" measurement to determine how to iterate through their pixels.

Reply:
In draw.c you can find the example of how to iterate pixels for single y value: https://github.com/pygame/pygame/blob/main/src_c/draw.c#L1357